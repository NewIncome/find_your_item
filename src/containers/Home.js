import React from 'react';
import { connect } from 'react-redux';
import { useHistory } from 'react-router-dom';
import PropTypes from 'prop-types';
import { addUsername } from '../actions';

const Home = props => {
  const { username, addUsername } = props;
  const history = useHistory();
  // const flag = 'ok';

  const onInput = event => addUsername(event.target.value);

  const onSubmit = event => {
    event.preventDefault();

    // if (username === '') {

    // } else if (username.length < 4) {

    // } else return

    history.push('/user');
  };

  return (
    <section className="section" id="Home">
      <h2>Find Your Item</h2>
      <form>
        <input
          placeholder="Please input your User Name"
          value={username}
          onChange={onInput}
        />
        <button
          type="submit"
          onClick={onSubmit}
        >
          Enter
        </button>
      </form>
    </section>
  );
};

Home.propTypes = {
  username: PropTypes.string.isRequired,
  addUsername: PropTypes.func.isRequired,
};

const mapStateToProps = username => username;

const mapDispatchToProps = dispatch => ({
  addUsername: username => dispatch(addUsername(username)),
});

export default connect(mapStateToProps, mapDispatchToProps)(Home);
