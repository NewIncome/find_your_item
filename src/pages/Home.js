import React, { useEffect, useState } from 'react';
import { connect } from 'react-redux';
import { useHistory } from 'react-router-dom';
import { bindActionCreators } from 'redux';
import PropTypes from 'prop-types';
import Error from '../components/Error';
import Loader from '../components/Loader';
import * as MyActions from '../actions';

const Home = props => {
  const {
    actions,
    user,
    fetchCall,
    username,
  } = props;
  const [loginFlag, setLoginFlag] = useState(true);
  const history = useHistory();
  const loginUrl = 'https://findmyitem-api.herokuapp.com/login';
  const signupUrl = 'https://findmyitem-api.herokuapp.com/users';

  const onInput = event => actions.addUsername(event.target.value);

  const onNewSubmit = event => {
    event.preventDefault();
    actions.fetchAPIcall(signupUrl, 'post', { data: { payload: { user: { name: username } } } });
  };

  const onLogSubmit = event => {
    event.preventDefault();
    actions.fetchAPIcall(loginUrl, 'post', { name: username });
  };

  const onClick = () => setLoginFlag(false);

  useEffect(() => {
    if (fetchCall.apiData) actions.setUserInfo(fetchCall.apiData);
  });

  useEffect(() => {
    if (username === user.name) {
      actions.fetchAPIreset();
      history.push('/user');
    }
  }, [user]);

  return (
    <>
      <section className="home-section" id="Home">
        <h2 className="home-h h">Find Your Item</h2>
        {fetchCall.loading
          ? <Loader />
          : (
            <>
              {/* eslint-disable react/jsx-one-expression-per-line */}
              <p className="t-text p">Hi there! {loginFlag ? 'Login' : 'Signup'} to start looking at items</p>
              <form className="home-form">
                <input
                  className="home-input"
                  placeholder="Please input your User Name"
                  value={username}
                  onChange={onInput}
                />
                {fetchCall.error ? <Error /> : ''}
                {loginFlag
                  ? (
                    <button className="home-btn" type="submit" onClick={onLogSubmit}>
                      <b>Login User</b>
                    </button>
                  )
                  : (
                    <button className="home-btn" type="submit" onClick={onNewSubmit}>
                      <b>Signup User</b>
                    </button>
                  )}
              </form>
              {loginFlag
                ? (
                  <button
                    className="signup"
                    type="button"
                    onClick={onClick}
                  >
                    Not signed up?
                  </button>
                ) : ''}
            </>
          )}
        <br />
      </section>
    </>
  );
};

Home.propTypes = {
  actions: PropTypes.objectOf(PropTypes.any).isRequired,
  user: PropTypes.objectOf(PropTypes.any).isRequired,
  fetchCall: PropTypes.objectOf(PropTypes.any).isRequired,
  username: PropTypes.string.isRequired,
};

const mapStateToProps = ({ username, user, fetchCall }) => ({ username, user, fetchCall });

function mapActionsToProps(dispatch) {
  return {
    actions: bindActionCreators({ ...MyActions }, dispatch),
  };
}

export default connect(mapStateToProps, mapActionsToProps)(Home);
