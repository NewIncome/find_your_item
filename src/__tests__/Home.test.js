import React from 'react';
import { Provider } from 'react-redux';
import renderer from 'react-test-renderer';
import configureStore from 'redux-mock-store';
import '@testing-library/jest-dom';
import Home from '../pages/Home';

const state = {
  username: 'test',
  user: {
    created_at: '2020-02-20T00:20:20.824Z',
    id: 20,
    name: 'Dummy User',
    updated_at: '2020-02-20T00:20:20.824Z',
  },
  fetchCall: { loading: false, error: null, apiData: null },
};

const mockStore = configureStore([]);

describe('My connected Home component', () => {
  let store;
  let component;

  beforeEach(() => {
    store = mockStore({ ...state });

    component = renderer.create(
      <Provider store={store}>
        <Home />
      </Provider>,
    );
  });

  it('is displayed correctly', () => {
    expect(component.toJSON()).toMatchSnapshot();
  });

  it("has a 'h2' title", () => {
    expect(component.root.findByType('h2').children).toStrictEqual(['Find Your Item']);
  });

  it("has an 'input' field", () => {
    expect(component.root.findByType('input').children).toStrictEqual([]);
  });
});
