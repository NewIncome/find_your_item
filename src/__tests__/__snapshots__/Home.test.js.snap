// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`My connected Home component is displayed correctly 1`] = `
<section
  className="home-section"
  id="Home"
>
  <h2
    className="home-h h"
  >
    Find Your Item
  </h2>
  <p
    className="t-text p"
  >
    Hi there! 
    Login
     to start looking at items
  </p>
  <form
    className="home-form"
  >
    <input
      className="home-input"
      onChange={[Function]}
      placeholder="Please input your User Name"
      value="test"
    />
    
    <button
      className="home-btn"
      onClick={[Function]}
      type="submit"
    >
      <b>
        Login User
      </b>
    </button>
  </form>
  <button
    className="signup"
    onClick={[Function]}
    type="button"
  >
    Not signed up?
  </button>
  <br />
</section>
`;
